From: Mike Kasick <mike@kasick.org>
Date: Tue, 18 Oct 2011 21:29:44 -0400

Enable caching and buffering for all of physical RAM in the decompressor.

Old method is to enable caching and buffering only for the 256 MB at the
start of the decompressor image.  This makes a hardboot-kexec'd kernel very
slow to decompress since the decompressor is located far above the kernel
destination.  This patch reduces boot time from 35 to 8 seconds.

diff --git a/kernel/arch/arm/boot/compressed/head.S b/kernel/arch/arm/boot/compressed/head.S
--- a/kernel/arch/arm/boot/compressed/head.S
+++ b/kernel/arch/arm/boot/compressed/head.S
@@ -9,6 +9,7 @@
  * published by the Free Software Foundation.
  */
 #include <linux/linkage.h>
+#include <asm/memory.h>
 
 /*
  * Debugging stuff
@@ -413,9 +414,14 @@
  * bits for the RAM area only.
  */
 		mov	r0, r3
+#if defined(PHYS_OFFSET) && defined(END_MEM)
+		mov	r9, #PHYS_OFFSET	@ start of RAM
+		mov	r10, #END_MEM		@ end of RAM
+#else
 		mov	r9, r0, lsr #18
 		mov	r9, r9, lsl #18		@ start of RAM
 		add	r10, r9, #0x10000000	@ a reasonable RAM size
+#endif
 		mov	r1, #0x12
 		orr	r1, r1, #3 << 10
 		add	r2, r3, #16384
diff --git a/kernel/arch/arm/mach-s5pv210/include/mach/memory.h b/kernel/arch/arm/mach-s5pv210/include/mach/memory.h
--- a/kernel/arch/arm/mach-s5pv210/include/mach/memory.h
+++ b/kernel/arch/arm/mach-s5pv210/include/mach/memory.h
@@ -17,6 +17,9 @@
 #define PHYS_OFFSET		UL(0x20000000)
 #else
 #define PHYS_OFFSET		UL(0x30000000)
+#ifdef CONFIG_DDR_RAM_3G
+#define END_MEM			UL(0x58000000)
+#endif
 #endif
 
 #define CONSISTENT_DMA_SIZE	(SZ_8M + SZ_4M + SZ_2M)
